### PR TITLE
Normalise PCA space

### PR DESCRIPTION
**Pull request by th0rntwig**
Normalisation of principal component transformed features.

PCA can be enabled by the user to reduce the feature space by finding the components that describe 99.9% of the variance in the dataset. This transformation is done on normalised [-1,1] data but the transformed data is no longer normalised to [-1,1]. This PR enables normalisation of the transformed data, and ensures that transformed test and prediction data is normalised to the corresponding transformed training data.

### CODE REVIEW

File: freqtrade/freqai/data_kitchen.py [hunk 1]
```diff
@@ -287,19 +287,26 @@ def normalize_data(self, data_dictionary: Dict) -> Dict[Any, Any]:
...
 
         for item in train_max.keys():
-            self.data[item + "_max"] = train_max[item]
-            self.data[item + "_min"] = train_min[item]
+            if not [col for col in df_train_features.columns if col.startswith('PC')]:
+                self.data[item + "_max"] = train_max[item]
+                self.data[item + "_min"] = train_min[item]
+            else:
+                # if PCA is enabled and has transformed the training features
+                self.data[item + "_pca_max"] = train_max[item]
```
Comment (robcaulk) [inline freqtrade/freqai/data_kitchen.py]: `item` here is going to be unique already. In the case of principal components, `item` will be `PC1` `PC2` etc. Can't we just leave this code without the check then since the keys will be unique already in the coin meta data (`self.data`)?

Or is there another reason why you attach `_pca_max`, and I am missing it?

File: freqtrade/freqai/data_kitchen.py [hunk 2]
```diff
@@ -320,8 +327,14 @@ def normalize_data(self, data_dictionary: Dict) -> Dict[Any, Any]:
                     - 1
                 )
 
-            self.data[f"{item}_max"] = train_labels_max  # .to_dict()
-            self.data[f"{item}_min"] = train_labels_min  # .to_dict()
+            if not [col for col in df_train_features.columns if col.startswith('PC')]:
```
Comment (robcaulk) [inline freqtrade/freqai/data_kitchen.py]: Same as comment above, I dont think we need this check since `item` is unique already wrt principal components or normal features.

File: freqtrade/freqai/data_kitchen.py [hunk 3]
```diff
@@ -331,11 +344,17 @@ def normalize_data_from_metadata(self, df: DataFrame) -> DataFrame:
         :param df: Dataframe to be standardized
         """
 
+        if not [col for col in df.columns if col.startswith('PC')]:
```
Comment (robcaulk) [inline freqtrade/freqai/data_kitchen.py]: If we are lucky, even this function will remain entirely unchanged since `item` is unique. See above comments and please confirm my suspicion.

File: freqtrade/freqai/data_kitchen.py [hunk 4]
```diff
@@ -450,38 +469,42 @@ def principal_component_analysis(self) -> None:
 
         from sklearn.decomposition import PCA  # avoid importing if we dont need it
 
-        n_components = self.data_dictionary["train_features"].shape[1]
-        pca = PCA(n_components=n_components)
+        pca = PCA(0.999)
```
Comment (robcaulk) [inline freqtrade/freqai/data_kitchen.py:468]: Nice this is more computationally efficient compared to stable.

File: freqtrade/freqai/data_kitchen.py [hunk 5]
```diff
@@ -450,38 +469,42 @@ def principal_component_analysis(self) -> None:
...
-        n_keep_components = np.argmin(pca.explained_variance_ratio_.cumsum() < 0.999)
-        pca2 = PCA(n_components=n_keep_components)
+        n_keep_components = pca.n_components_
         self.data["n_kept_components"] = n_keep_components
-        pca2 = pca2.fit(self.data_dictionary["train_features"])
+        n_components = self.data_dictionary["train_features"].shape[1]
         logger.info("reduced feature dimension by %s", n_components - n_keep_components)
-        logger.info("explained variance %f", np.sum(pca2.explained_variance_ratio_))
-        train_components = pca2.transform(self.data_dictionary["train_features"])
+        logger.info("explained variance %f", np.sum(pca.explained_variance_ratio_))
```
Comment (robcaulk) [inline freqtrade/freqai/data_kitchen.py:474]: I know the old code used these strings, but in general we prefer the use of f-strings. So any chance we touch some old code that doesnt use f-strings, ideally we would convert it. I can do that on this branch. Just a comment for future PRs.

File: freqtrade/freqai/data_kitchen.py [hunk 6]
```diff
@@ -450,38 +469,42 @@ def principal_component_analysis(self) -> None:
...
         if self.freqai_config.get('data_split_parameters', {}).get('test_size', 0.1) != 0:
-            test_components = pca2.transform(self.data_dictionary["test_features"])
+            test_components = pca.transform(self.data_dictionary["test_features"])
             self.data_dictionary["test_features"] = pd.DataFrame(
                 data=test_components,
                 columns=["PC" + str(i) for i in range(0, n_keep_components)],
                 index=self.data_dictionary["test_features"].index,
             )
+            # normalise transformed test feature to transformed training features
+            self.data_dictionary["test_features"] = self.normalize_data_from_metadata(
```
Comment (robcaulk) [inline freqtrade/freqai/data_kitchen.py:499]: This is an interesting component of this PR that I hope has a meaningful impact on the principal component feature. Thanks for this.

**Review by robcaulk — CHANGES_REQUESTED** [review]
Generally speaking - this is going to improve principal component analysis feature. 

But I think we get some of this already for "free" without adding all these extra strings to the metadata. See my comment and let me know if I missed something.

File: freqtrade/freqai/data_kitchen.py (hunk 1 above)
Comment (th0rntwig) [reply freqtrade/freqai/data_kitchen.py]: You're right, the keys are already unique, my mistake!

File: freqtrade/freqai/data_kitchen.py (hunk 4 above)
Comment (th0rntwig) [reply freqtrade/freqai/data_kitchen.py:468]: And more readable imo :)

File: freqtrade/freqai/data_kitchen.py (hunk 5 above)
Comment (th0rntwig) [reply freqtrade/freqai/data_kitchen.py:474]: Good to know, will do that from now on!

File: freqtrade/freqai/data_kitchen.py (hunk 6 above)
Comment (th0rntwig) [reply freqtrade/freqai/data_kitchen.py:499]: I think it will improve the performance of the models when the user enables PCA. Looking forward to seeing if that's correct.

**Review by robcaulk — APPROVED** [review]
